### ISSUE TITLE
Vers 2.1.1 in TB91: account order does not survive restart (ExQuilla account)

### ISSUE THREAD

**Issue by 2019-05-10** [protz/Manually-Sort-Folders#170]
TB 91.2.1 (64-bit)
Manually-Sort-Folders 2.1.1

Accounts can be orderd, but as soon as TB is restarted the order is lost again.

**Comment by pstein** [comment]
I can confirm the problems for TB v91.4.1 (on windows)
After some time or some other (currently unknown) action the sort order is scrambled.
When I visit again "Manually sort folders" AddOn settings the sort order here is still valid.
When I click the button "Refresh folder pane" then everything is ok again.

But having to do it is annoying.

In my case restart Thunderbird is not the problem. The folder order seems to survive a restart - at least here.
So it must be another reason

**Comment by pancrazio111** [comment]
Same problem here
Version , 2.1.1, Thunderbird Windows 91.5
can sort all account except for the Exquilla one with last TB update - when sorted, it reverts back when I restart TB

**Comment by itagagaki** [comment]
I want to investigate this issue, but I don't have an Exchange email account, and ExQuilla is a paid add-on, so I can't try it.
I'm not sure how I can.

**Comment by pstein** [comment]
Just to clarify:

The order of folders is reset only occasionally and only for some accounts.

I am not using an "ExQuilla or Exchange" software

Maybe some other AddOns or certain TB preference is confusing Manually-Sort-Folders.

Maybe a Logging for Manually-Sort-Folders would be useful.
Whenever a Manually-Sort-Folders Trigger detects changed sequence (e.g. at startup) a popup should appear.

**Comment by itagagaki** [comment]
@pstein Hmm, so it's not just a problem with ExQuilla or Exchange.
Thank you for the report.

**Comment by itagagaki** [comment]
@pltein First, this issue reported by @2019-05-10 and @pancrazio111 is about the order of accounts, but maybe you are reporting about the order of folders in each account?
And this issue is about the ExQuilla account, so if it is not related to that, you can report it as a new issue.

**Comment by itagagaki** [comment]
I was able to install ExQuilla and add an Exchange account.
ExQuilla has a free trial period of one month.
During which period we have to resolve this issue.

I tried with the following set, and I was able to change the order of the Exchange accounts without any problem, and the changes were retained even after restarting Thunderbird. I mean that I couldn't reproduce this issue.

Windows 10 Pro 21H1
Thunderbird 91.5.0 (64-bit) (ja)
ExQuilla 91.1
Manually Sort Folders 2.1.1

Could you please check if the issue still exists?

**Comment by pancrazio111** [comment]
yes, still there in Win10 home 21h2, tb 91.5.0-64bit, exquilla 91.1
(this doesn't happen with other accounts, where changes are maintained upon restart)
be aware if you added the exquilla acct to your tb it will probably put the last acct you added on 
top of the others.
p

**Comment by itagagaki** [comment]
It doesn't reproduce on my side, so I need a clue to know what's going on.
Launch Thunderbird and press Shift+Ctrl+J to bring up the error console.
Extract the lines containing tbsortfolders and exquilla from there and let me know. (Please hide the parts that you want.)

Here's what I get.

> 1642863459117	tbsortfolders.folderPane	DEBUG	Init
1642863459117	tbsortfolders.folderPane	DEBUG	Add observer
2022-01-22 23:57:39.145	exquilla.base	INFO	Adding pre-existing exquilla account account29
2022-01-22 23:57:39.261	exquilla.base	INFO	manageNtlmUri Adding Spec https://\*\*\*\*.\*\*\*\*\*\*\*.\*\*/\*\*\*/\*\*\*\*\*\*\*\*.\*\*\*\* host https://\*\*\*\*.\*\*\*\*\*\*\*.\*\*
2022-01-22 23:57:39.704	exquilla.native	INFO	Creating a new mailbox with URI exquilla://\*\*\*\*\*\*\*\*%2E\*\*\*\*\*\*\*%40\*\*\*\*\*\*\*%2E\*\*\@\*\*\*\*.\*\*\*\*\*\*\*.\*\*
2022-01-22 23:57:42.585	exquilla.base	INFO	Starting EwsNotification
2022-01-22 23:57:42.587	exquilla.base	INFO	New Address Book with URI exquilla-directory://\*\*\*\*\*\*\*\*%2E\*\*\*\*\*\*\*%40\*\*\*\*\*\*\*%2E\*\*\@\*\*\*\*.\*\*\*\*\*\*\*.\*\*/%23aH
2022-01-22 23:57:42.591	exquilla.base	INFO	New Address Book with URI exquilla-directory://\*\*\*\*\*\*\*\*%2E\*\*\*\*\*\*\*%40\*\*\*\*\*\*\*%2E\*\*\@\*\*\*\*.\*\*\*\*\*\*\*.\*\*/%23aH/%7B06967759-274D-40B2-A3EB-D7F9E73727D7%7D
2022-01-22 23:57:42.594	exquilla.base	INFO	New Address Book with URI exquilla-directory://\*\*\*\*\*\*\*\*%2E\*\*\*\*\*\*\*%40\*\*\*\*\*\*\*%2E\*\*\@\*\*\*\*.\*\*\*\*\*\*\*.\*\*/%23aH/%7BA9E2BC46-B3A0-4243-B315-60D991004455%7D
2022-01-22 23:57:42.597	exquilla.base	INFO	New Address Book with URI exquilla-directory://\*\*\*\*\*\*\*\*%2E\*\*\*\*\*\*\*%40\*\*\*\*\*\*\*%2E\*\*\@\*\*\*\*.\*\*\*\*\*\*\*.\*\*/%23aH/GAL%20Contacts
2022-01-22 23:57:42.600	exquilla.base	INFO	New Address Book with URI exquilla-directory://\*\*\*\*\*\*\*\*%2E\*\*\*\*\*\*\*%40\*\*\*\*\*\*\*%2E\*\*\@\*\*\*\*.\*\*\*\*\*\*\*.\*\*/%23aH/Organizational%20Contacts
2022-01-22 23:57:42.604	exquilla.base	INFO	New Address Book with URI exquilla-directory://\*\*\*\*\*\*\*\*%2E\*\*\*\*\*\*\*%40\*\*\*\*\*\*\*%2E\*\*\@\*\*\*\*.\*\*\*\*\*\*\*.\*\*/%23aH/PeopleCentricConversation%20Buddies
2022-01-22 23:57:42.608	exquilla.base	INFO	New Address Book with URI exquilla-directory://\*\*\*\*\*\*\*\*%2E\*\*\*\*\*\*\*%40\*\*\*\*\*\*\*%2E\*\*\@\*\*\*\*.\*\*\*\*\*\*\*.\*\*/%23aH/Recipient%20Cache
2022-01-22 23:57:42.612	exquilla.base	INFO	New Address Book with URI exquilla-directory://\*\*\*\*\*\*\*\*%2E\*\*\*\*\*\*\*%40\*\*\*\*\*\*\*%2E\*\*\@\*\*\*\*.\*\*\*\*\*\*\*.\*\*/%23aH/%1A%3E
2022-01-22 23:57:42.616	exquilla.base	INFO	New Address Book with URI exquilla-directory://\*\*\*\*\*\*\*\*%2E\*\*\*\*\*\*\*%40\*\*\*\*\*\*\*%2E\*\*\@\*\*\*\*.\*\*\*\*\*\*\*.\*\*/GAL

Also, please let me know if there are any error messages or exceptions in the error console that concern you.

**Comment by itagagaki** [comment]
@pancrazio111 @2019-05-10 Don't you have any additional info?

**Comment by itagagaki** [comment]
If you do not cooperate with me in the next ten days, I may not solve the issue
because my free trial period of Exquilla will be expired.

**Comment by pancrazio111** [comment]
hi sorry I had responded by email but probably need to do it here as looks like you din't get me email
this is the error console

Extract the lines containing tbsortfolders and exquilla from there and let me know. (Please hide the parts that you want.)

Here's what I get.

    1642863459117 tbsortfolders.folderPane DEBUG Init
    1642863459117 tbsortfolders.folderPane DEBUG Add observer
    2022-01-22 23:57:39.145 exquilla.base INFO Adding pre-existing exquilla account account29
    2022-01-22 23:57:39.261 exquilla.base INFO manageNtlmUri Adding Spec https://****.*******.**/***/********.**** host https://****.*******.**
    2022-01-22 23:57:39.704 exquilla.native INFO Creating a new mailbox with URI exquilla://********%2E*******%40*******%2E**@****.*******.**
    2022-01-22 23:57:42.585 exquilla.base INFO Starting EwsNotification
    2022-01-22 23:57:42.587 exquilla.base INFO New Address Book with URI exquilla-directory://********%2E*******%40*******%2E**@****.*******.**/%23aH
    2022-01-22 23:57:42.591 exquilla.base INFO New Address Book with URI exquilla-directory://********%2E*******%40*******%2E**@****.*******.**/%23aH/%7B06967759-274D-40B2-A3EB-D7F9E73727D7%7D
    2022-01-22 23:57:42.594 exquilla.base INFO New Address Book with URI exquilla-directory://********%2E*******%40*******%2E**@****.*******.**/%23aH/%7BA9E2BC46-B3A0-4243-B315-60D991004455%7D
    2022-01-22 23:57:42.597 exquilla.base INFO New Address Book with URI exquilla-directory://********%2E*******%40*******%2E**@****.*******.**/%23aH/GAL%20Contacts
    2022-01-22 23:57:42.600 exquilla.base INFO New Address Book with URI exquilla-directory://********%2E*******%40*******%2E**@****.*******.**/%23aH/Organizational%20Contacts
    2022-01-22 23:57:42.604 exquilla.base INFO New Address Book with URI exquilla-directory://********%2E*******%40*******%2E**@****.*******.**/%23aH/PeopleCentricConversation%20Buddies
    2022-01-22 23:57:42.608 exquilla.base INFO New Address Book with URI exquilla-directory://********%2E*******%40*******%2E**@****.*******.**/%23aH/Recipient%20Cache
    2022-01-22 23:57:42.612 exquilla.base INFO New Address Book with URI exquilla-directory://********%2E*******%40*******%2E**@****.*******.**/%23aH/%1A%3E
    2022-01-22 23:57:42.616 exquilla.base INFO New Address Book with URI exquilla-directory://********%2E*******%40*******%2E**@****.*******.**/GAL

**Comment by pancrazio111** [comment]
also TB now update to 91.5.1 and exquilla yesterday to 91.2 but still no luck!

**Comment by itagagaki** [comment]
@pancrazio111 Thank you. But the one you pasted looks like my log.
It is exactly the same and the date is 2022-01-22.
Could you please check again?

**Comment by pancrazio111** [comment]
I am sorry to fthe confusion -
here it is

18:18:48.091 Unexpected event profile-after-change URLQueryStrippingListService.jsm:224

18:18:49.531 Unknown Collection "thunderbird/query-stripping" RemoteSettingsClient.jsm:160

18:18:51.001 1642871931000               tbsortfolders.folderPane               DEBUG Init

18:18:51.001 1642871931001               tbsortfolders.folderPane               DEBUG Add observer

18:18:51.067 2022-01-22 18:18:51.067               exquilla.base     ERROR  Test error logging: Error(chrome://exquilla/content/exquilla.js:115:18) JS Stack trace: [onLoad@exquilla.js:115:18](mailto:onLoad@exquilla.js:115:18)

@exquilla.js:1012:56 Log.jsm:723

18:18:51.068 2022-01-22 18:18:51.068               exquilla.base     INFO     Adding pre-existing exquilla account account16

18:18:51.139 2022-01-22 18:18:51.139               exquilla.base     INFO               manageNtlmUri Adding Spec https://autodiscover.xxx.uk/EWS/Exchange.asmx host https://autodiscover.xxx.uk/

18:18:52.249 tb.account.size_on_disk - Truncating float/double number. 12

18:18:52.250 tb.account.size_on_disk - Attempted to set the scalar to an incompatible value. 6

18:18:52.250 tb.account.size_on_disk - Truncating float/double number. 4

18:18:52.250 tb.account.size_on_disk - Trying to set an unsigned scalar to a negative number. 8

18:18:52.334 NS_ERROR_FILE_CORRUPTED: Component returned failure code: 0x8052000b (NS_ERROR_FILE_CORRUPTED) [mozIStorageService.openDatabase] CalStorageCalendar.jsm:210

**Comment by itagagaki** [comment]
I don't see any suspicious output in MSF or in ExQuilla.
However, ExQuilla's output should continue further, but if you don't see that output, then ExQuilla may be stuck in the middle.
The last five lines suggest that there is something wrong with Thunderbird itself. The last line, in particular, is clearly an error indicating that the file is corrupt and failing to load.

Can you please try the following?
Start Thunderbird with the -p option. For example, if `thunderbird.exe` is located in `C:\Program Files\Mozilla Thunderbird`, then at the command prompt, type
```
"C:\Program Files\Mozilla Thunderbird\thunderbird.exe" -p
```
at the command prompt.
This will bring up a dialog window asking you to select a user profile, so create a new profile and start Thunderbird with it. Then try adding regular mail accounts and/or ExQuilla accounts. Then check the behavior of MSF and ExQuilla.

**Comment by phantomtaint** [comment]
I'm seeing the following when manually changing the order of my accounts (moving Exquilla "account5" to the top).

Could the write warning be related to the sort order not persisting after Thunderbird closes?

`1644512803377    tbsortfolders.ui    DEBUG    on_account_move_up
1644512803377    tbsortfolders.ui    DEBUG    index: 2
1644512803378    tbsortfolders.ui    DEBUG    Sorted accounts: account2,account5,account3
1644512803591    tbsortfolders.ui    DEBUG    on_account_move_up
1644512803591    tbsortfolders.ui    DEBUG    index: 1
1644512803592    tbsortfolders.ui    DEBUG    Sorted accounts: account5,account2,account3
Warning: attempting to write 568049 bytes to preference [extensions.tbsortfolders@xulforum.org.tbsf_data](mailto:extensions.tbsortfolders@xulforum.org.tbsf_data). This is bad for general performance and memory usage. Such an amount of data should rather be written to an external file. This preference will not be sent to any content processes.`

**Comment by itagagaki** [comment]
If the order is lost when you restart Thunderbird, it may be that the order is not lost internally, but the folder tree view pane is not updated correctly.

Please wait for the improved version 2.1.2. It should be available on ATN soon. You can also install Release version 2.1.2 from this repository if you want to try it out right away.

Regarding the warning @phantomtaint saw, if the prefs.js file is not corrupt, the total number of folders is too large. Its size seems to be an order of magnitude too large. Open the pref.js file of the profile you are using (you can find it in Help->Troubleshooting Information) with a text editor or something, and search `extensions.tbsortfolders@xulforum.org.tbsf_data` and see if there is anything wrong with that line. However, this is only a warning. So it is unlikely to be the cause of the problem directly.

**Comment by pancrazio111** [comment]
I will try, thanks, and wait for 2.1.2. Also, I have a second pc at work, intalled MSF and works well - same set up. So it may be something with my laptop's TB. I am not sure I would try fixes as I don't want to go through all the nightmare of reconfiguring exquilla - it's not an easy thing...

**Comment by phantomtaint** [comment]
> Regarding the warning @phantomtaint saw, if the prefs.js file is not corrupt, the total number of folders is too large. Its size seems to be an order of magnitude too large. Open the pref.js file of the profile you are using (you can find it in Help->Troubleshooting Information) with a text editor or something, and search `extensions.tbsortfolders@xulforum.org.tbsf_data` and see if there is anything wrong with that line. However, this is only a warning. So it is unlikely to be the cause of the problem directly.

My folder tree is indeed unavoidably large.

The tbsf_data line in prefs.js is intact.  It's gigantic, but it's intact.

I'll look out for the new addon version, thanks.

**Comment by itagagaki** [comment]
I thought this issue was related to ExQuilla, but maybe it was not.
Anyway, MSF 2.1.2 is now available at addons.thunderbird.net. Please try it out.

**Comment by phantomtaint** [comment]
Unfortunately the account order reverting after closing Thunderbird still persists for me on version 2.1.2.

**Comment by itagagaki** [comment]
@phantomtaint That's too bad.
Confirmation. You are referring to the order of accounts, not the order of folders, right?
Did you know that Thunderbird 91 has an account reordering feature of its own?
What happens if you temporarily disable MSF and use the Thunderbird feature to reorder the accounts and try that?

If you are talking about the order of the folders, could you try the following?
When you restart Thunderbird, if you find that the order of the folders has reverted, try opening MSF, going to the [Sort folders] tab, and clicking [Refresh folder pane].

**Comment by phantomtaint** [comment]
> You are referring to the order of accounts, not the order of folders, right?

Yes, I only use the addon to change the order of accounts, not folders.

The built in account sorting function in Thunderbird doesn't work at all for forcing the Exquilla account to the top (it will let me move it, but it doesn't actually do anything).  The MSF addon does actually let me move the Exquilla account to the top, at least until Thunderbird is closed and reopened.  I've tried the built in Thunderbird sorting with and without MSF enabled, no difference.

**Comment by itagagaki** [comment]
@phantomtaint I understand how things are going on your end.
Both Thunderbird's built-in function and MSF are working fine on my end, even for an ExQuilla account.

please try to confirm the following?

1. Enable MSF
2. Restart Thunderbird
3. Press `Shift+Ctrl+J` to open Error Console
4. Open MSF
5. You should see the following log output in the error console. See if the account information printed here is in the order you expect, or if it is still back to the way it was before the reordering.

```
1644985150133	tbsortfolders.ui	DEBUG	on_load
1644985150133	tbsortfolders.ui	DEBUG	Total accounts: NNN
1644985150133	tbsortfolders.ui	DEBUG	Account: ***
1644985150134	tbsortfolders.ui	DEBUG	Account: ***
1644985150134	tbsortfolders.ui	DEBUG	Account: ***
```

**Comment by phantomtaint** [comment]
I discovered today that the manual sorting of accounts built into Thunderbird IS now working, but it also reverts order after Thunderbird restarts, so the same result as the MSF addon.  I guess one of the recent updates changed something.

I did as you requested and observed the output in the console.  The account list matches the same reverted order shown in the folder pane, with Local Folders at the top instead of Exquilla.

Clicking 'Refresh folder pane' doesn't change anything, which isn't surprising based on the displayed account order in the console.

Since the manual sort function in Thunderbird is also reverting on restart, it looks like Thunderbird itself may be to blame.

**Comment by itagagaki** [comment]
One question: does this happen even if you disable all add-ons except ExQuilla?

**Comment by pancrazio111** [comment]
yes - even if I disable MSF

**Comment by phantomtaint** [comment]
Yes, just tested.  Disabling all addons except Exquilla has no effect.

**Comment by itagagaki** [comment]
Let's monitor the prefs.js file for changes.
First, download and launch the baretail.
This is a text file viewer that can monitoring in real-time.
[Bare Metal Software &gt; BareTail - Free tail for Windows](http://www.baremetalsoft.com/baretail/)

Open your prefs.js file with baretail.
you can find the file in Thunderbird Help->Troubleshooting Information

Find a line contains `"mail.accountmanager.accounts"`
Unfortunately, baretail doesn't have a search function, but the contents of prefs.js are probably in alphabetical order, so it shouldn't be too hard to find.

This line is the order of the accounts. Let's see at what point this changes.
Start Thunderbird and try to sort your accounts. You should be able to see the change in baretail.
And keep an eye on it to see when it changes against your will. You won't know who made the change, but if you know the timing, it will give you a clue.
You may also want to open the error console of Thunderbird to get more clues.

**Comment by phantomtaint** [comment]
I did find something interesting while monitoring this setting in the prefs.js.

Right when Thunderbird starts, before the application window is even visible, the Exquilla account in this line is removed, and then a second later it is added back (and appended to the end instead of maintaining it's postion).

Setting when accounts are ordered as desired with Exquilla (account5) at the top of the folder pane:

```user_pref("mail.accountmanager.accounts", "account5,account2,account3");```

Setting immediately after launching Thunderbird:

```user_pref("mail.accountmanager.accounts", "account2,account3");```

Setting a second later:

```user_pref("mail.accountmanager.accounts", "account2,account3,account5");```

The order is always maintained after Thunderbird closes, only reverting when it is reopened.  This occurs with or without MSF being enabled.

Perhaps it would be possible for the MSF addon to monitor this line and revert it after a few seconds of initialization if it detects the same accounts existing in a different order.

**Comment by phantomtaint** [comment]
If I disable Exquilla, the account entry in the mail.accountmanager.accounts config still disappears when Thunderbird starts up, and at that point it does not get added back until the Exquilla addon is re-enabled.  When I toggle Exquilla back on, the account entry is immediately added back in the config at the end of the list.

**Comment by itagagaki** [comment]
I guess we can assume that Thunderbird erases account5 on startup, and ExQuilla restores it to the end again.
I think it will be difficult for MSF to solve that.
But if I come up with something, I'll write again.

Since it is a behavioral matter of Thunderbird, you may get something if you report your investigation to
https://bugzilla.mozilla.org/

**Comment by itagagaki** [comment]
The same thing happened to me. I don't know why it's happening when it never happened before.
But it's good news. I'm not sure if I can figure it out, but now I have a sample. I'll investigate this issue again.

**Comment by itagagaki** [comment]
@phantomtaint I have one confirmation for you.
Are you continuing to pay Beonex and is your ExQuilla license still active?
Because I only reproduced the issue after my ExQuilla trial period was over.

**Comment by phantomtaint** [comment]
@itagagaki Yes, I have an active Exquilla license.

**Comment by phantomtaint** [comment]
@itagagaki After changing the account type from "exquilla" to "none" and reopening Thunderbird, Thunderbird froze up.  It was using a small amount of cpu and 1GB of ram but never loaded the folder pane, message pane etc.  I let it sit there for 5 minutes before force closing it.  I changed the type back to "exquilla", restarted and sure enough that account stays on top now.  I've closed and reopened multiple times and it is no longer moving down.

I don't understand it either, but that did the trick.
